src/lib/fs/path.c

```diff
@@ -293,3 +341,389 @@ make_path_absolute(char *fname)
+ * empty). Returns false otherwise. */
+static bool
+is_file(file_status_t file_type, const char *pattern_after)
+{
+  (void)pattern_after;
+  return file_type != FN_ERROR && file_type != FN_NOENT && file_type != FN_DIR;
+}
+
+/** Returns true if <b>file_type</b> represents an existing directory. Returns
+ * false otherwise. */
+static bool
+is_dir(file_status_t file_type, const char *pattern_after)
+{
+  (void)pattern_after;
+  return file_type == FN_DIR;
+}
+
+#ifdef _WIN32
+/** Returns true if the path where <b>file_type</b> was obtained from
+ * concatenated with <b>pattern_after</b> is a complete path and points to an
+ * existing file or directory. Returns false otherwise. */
+static bool
+no_pattern_after(file_status_t file_type, const char *pattern_after)
+{
+  return file_type != FN_ERROR && file_type != FN_NOENT && !*pattern_after;
+}
+
+/** Returns true if the path where <b>file_type</b> was obtained from points to
+ * an existing directory and <b>pattern_after</b> is not empty. */
+static bool
+dir_pattern_after(file_status_t file_type, const char *pattern_after)
+{
+  return is_dir(file_type, pattern_after) && *pattern_after;
+}
+#endif /* defined(_WIN32) */
+
+/*****************************************************************************/
+
+/** Returns a copy of <b>path</b> cut before <b>sep_index</b>. */
+static char *
+get_path_until_separator(const char *path, int sep_index)
+{
+  size_t len = sep_index < 1 ? sep_index + 1 : sep_index;
+  return tor_strndup(path, len);
+}
+
+/***** Helper functions of type handle_glob_fn used for glob processing ******/
+
+#ifdef _WIN32
+/** Returns true if <b>pattern</b> cut before <b>start_sep</b> is the path of
+ * an existing directory. Returns false otherwise. */
+static bool
+handle_glob_win32(const char *pattern, int start_sep, int end_sep,
+                  smartlist_t *result)
+{
+  (void)end_sep;
+  (void)result;
+
+  char *path_until_glob = get_path_until_separator(pattern, start_sep);
+  file_status_t file_type = file_status(path_until_glob);
+  tor_free(path_until_glob);
+  return file_type == FN_DIR;
+}
+#endif /* defined(_WIN32) */
+
+/** Adds <b>pattern</b> cut before <b>start_sep</b> to <b>results</b>. Returns
+ * true if <b>pattern</b> contains a glob after <b>end_sep</b>. Returns false
+ * otherwise. */
+static bool
+handle_glob_opened_files(const char *pattern, int start_sep, int end_sep,
+                         smartlist_t *result)
+{
+  char *path_until_glob = get_path_until_separator(pattern, start_sep);
+  smartlist_add(result, path_until_glob);
+  // if the following fragments have no globs, we're done
+  const char *pattern_after = &pattern[end_sep+1];
+  return has_glob(pattern_after);
+}
+
+/*****************************************************************************/
+
+#ifdef _WIN32
+/** Returns a list of paths that match <b>pattern</b>. Retruns NULL on error.
+ * Due to Windows API limitations, globs are only accepted on the last path
```

Comment:
Where is this limitation documented?

People might be surprised if they move working torrcs from other OSes to Windows, and they fail.

Reply:
Are there unit tests for globs on non-terminal parts of the path on Windows?
What happens?

Reply:
This will be documented and I will write an unit test to make sure there is an error instead of failing silently if a glob is used where it is not supported on Windows.

Reply:
I ended up not doing any refactoring as tor_listdir is in dir.c and unglob_file_name is in path.c and is a static function for that file. All the glob related functions in path.c would have to be moved to dir.c to do this without making a mess and I think path.c is the correct place for those functions.
I you think otherwise or have a better idea let me know.

Reply:
Can we make tor_listdir call a function in path.c that contains the common code?

Reply:
This is done in d401ac6.